### PR TITLE
feat: add metaworld environment for openpi models

### PR DESCRIPTION
**Pull request by chenkang455**
This PR adds MetaWorld environment support to the openpi models (pi0 and pi05). 

### Description

1. **MetaWorld Environment Implementation** (`rlinf/envs/metaworld/metaworld_env.py`):
   - Adapted from the Libero environment code structure, implemented the `MetaWorldEnv` class
   - Supports multi-task training on MT50 (50 tasks)

2. **Environment Vectorization Support**:
   - Implemented `ReconfigureSubprocEnv` class (modified from Libero version) supporting environment reconfiguration and autoreset mode
   - Supports using either `gym.vector.AsyncVectorEnv` or custom `ReconfigureSubprocEnv`
 
3. **Evaluation and Testing Tools**:
   - Implemented evaluation script (`metaworld_eval.py`) and test script (`metaworld_test.py`)
   - Added environment usage documentation (`rlinf/envs/metaworld/README.md`)
 
### Motivation and Context

This PR adds the MetaWorld environment for openpi models. Ref: https://metaworld.farama.org/

### How has this been tested?

Check `rlinf/envs/metaworld/README.md` for metaworld environment installation.

To run the RL on the pi0 model, run:
```
bash examples/embodiment/run_embodiment.sh metaworld_50_ppo_openpi
```

However, you have to manually add a novel openpi config in the openpi config code, which is not included in the rlinf code. 

```
TrainConfig(
    name="pi0_metaworld",
    model=pi0_config.Pi0Config(action_horizon=5),
    data=LeRobotMetaworldDataConfig(
        repo_id="lerobot/metaworld_mt50",
        base_config=DataConfig(
            prompt_from_task=True,
            root="data/metaworld_mt50"
        ),
        assets=AssetsConfig(
            assets_dir="checkpoints/torch/pi0_metaworld/assets"
        ),
        extra_delta_transform=False, # TODO: required
    ),
    weight_loader=weight_loaders.CheckpointWeightLoader("checkpoints/jax/pi0_base/params"),
    pytorch_weight_path="checkpoints/torch/pi0_base",
    num_train_steps=30_000,
),
```
Below is my training curve for the pi0 model with PPO:
image

### Additional information (optional, e.g., figures and logs):
I have some comments:
1. The main code of `metaworld_env.py` was adapted from the Libero environment. Would you consider implementing a base environment class so that other environments can inherit from it?
2. I modified a version of `ReconfigureSubprocEnv` (originally from Libero) to support autoreset mode. However, I noticed that this environment runs significantly slower than `gym.vector.AsyncVectorEnv` under the osmesa mode. In the `egl` mode, their performance is comparable. Besides, the `ReconfigureSubprocEnv` can be merged together?
3. Currently, the EGL rendering method is inelegantly set using `os.environ["MUJOCO_EGL_DEVICE_ID"] = str(self.seed_offset)`. Is there a better approach to managing this?
4. Autoreset mode is supported, but the value loss seems very strange; the provided config is set as the non-autoreset mode.
image

5. OpenPi model code in `src/openpi/training/config.py` needs to be adapted to the Metaworld environment, while the  `pi0_libero` and  `pi05_libero` configs are supported by the OpenPi official. We can consider placing it under the rlinf `rlinf/models/embodiment`

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/algorithms/losses.py [hunk 1]
```diff
@@ -168,10 +168,29 @@ def compute_ppo_critic_loss(
...
+
+    var_returns = torch.var(masked_returns)
+    if torch.isnan(var_returns) or var_returns == 0:
+        explained_variance = torch.tensor(float("nan"), device=returns.device)
+    else:
+        var_diff = torch.var(masked_returns - masked_values)
+        if torch.isnan(var_diff):
+            explained_variance = torch.tensor(float("nan"), device=returns.device)
+        else:
+            explained_variance = 1 - var_diff / var_returns
```
Comment (QuanluZhang) [inline rlinf/algorithms/losses.py:187]: @zanghz21 please help review this part

File: rlinf/envs/metaworld/README.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
+# Metaworld Environment Setup Guide
```
Comment (QuanluZhang) [inline rlinf/envs/metaworld/README.md:1]: This pr looks great! But as a user I don't know how to use it, the guideline is too concise. Could you add document for RL training on metaworld like the examples here https://rlinf.readthedocs.io/en/latest/rst_source/examples/index.html .

File: rlinf/envs/metaworld/metaworld_eval.py [hunk 3]
```diff
@@ -0,0 +1,190 @@
+# Copyright 2025 The RLinf Authors.
```
Comment (QuanluZhang) [inline rlinf/envs/metaworld/metaworld_eval.py]: this file should be placed under this folder: `https://github.com/RLinf/RLinf/tree/main/examples/embodiment` to align with other environments.

File: rlinf/envs/metaworld/metaworld_test.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
+# Copyright 2025 The RLinf Authors.
```
Comment (QuanluZhang) [inline rlinf/envs/metaworld/metaworld_test.py]: Could you please move this file to `https://github.com/RLinf/RLinf/tree/main/tests/e2e_tests/embodied` folder. You can follow the style of other environments' CI tests. CI test is very important to guarantee that this feature will not be broken by future PRs.

**Comment by QuanluZhang** [comment]
@chenkang455 This pr looks great! Together, let's futhur refine this pr to make it meet our high code/project quality standard.

File: rlinf/envs/metaworld/README.md (hunk 2 above)
Comment (chenkang455) [reply rlinf/envs/metaworld/README.md:1]: Yeah, i have supplemented the doc for metaworld in [docs/source-en/rst_source/examples/metaworld.rst](https://github.com/RLinf/RLinf/pull/242/files#diff-6fa12e1ce07786a3c71c7663834dc90f1c58d308117b9a599a7c397f9dad24dc)

File: rlinf/envs/metaworld/metaworld_eval.py (hunk 3 above)
Comment (chenkang455) [reply rlinf/envs/metaworld/metaworld_eval.py]: This is a simple code for eval on Metaworld with openpi. I have removed it for code consistency.

File: rlinf/envs/metaworld/metaworld_test.py (hunk 4 above)
Comment (chenkang455) [reply rlinf/envs/metaworld/metaworld_test.py]: Config for metaworld has been added in https://github.com/RLinf/RLinf/pull/242/files#diff-cb3cf6b640421398cfbb4266030a02990c60374edc6a185d68f08b7ce504eb21

**Comment by chenkang455** [comment]
Thanks for your time reviewing the codes, @QuanluZhang.

I have updated the documentation in https://github.com/RLinf/RLinf/pull/242/files#diff-cb3cf6b640421398cfbb4266030a02990c60374edc6a185d68f08b7ce504eb21.

**Comment by chenkang455** [comment]
Thanks for your time reviewing the codes @andylin-hao. After removing current MUJOCO_EGL_DEVICE_ID settings, the egl rendering seems all on the first device? After that, the gpu memory is oom.
image

Below is the right setting:
image

**Comment by andylin-hao** [comment]
> Thanks for your time reviewing the codes @andylin-hao. After removing current MUJOCO_EGL_DEVICE_ID settings, the egl rendering seems all on the first device? After that, the gpu memory is oom. image
> 
> Below is the right setting: image

Ah yes, sorry I wasn't very clear about this. My PR has not been merged yet, so you'll have to keep it for now. But you can write something like `os.environ["MUJOCO_EGL_DEVICE_ID"] = str(self._local_accelerator_id)`, which is a `Worker` member variable of the current assigned GPU. After my PR is merged, RLinf will automatically set this for every process so there's no need even for this. I'll let you know when the PR is merged (if your PR is merged first, I'll make the modification for you instead).

**Comment by chenkang455** [comment]
> > Thanks for your time reviewing the codes @andylin-hao. After removing current MUJOCO_EGL_DEVICE_ID settings, the egl rendering seems all on the first device? After that, the gpu memory is oom. image
> > Below is the right setting: image
> 
> Ah yes, sorry I wasn't very clear about this. My PR has not been merged yet, so you'll have to keep it for now. But you can write something like `os.environ["MUJOCO_EGL_DEVICE_ID"] = str(self._local_accelerator_id)`, which is a `Worker` member variable of the current assigned GPU. After my PR is merged, RLinf will automatically set this for every process so there's no need even for this. I'll let you know when the PR is merged (if your PR is merged first, I'll make the modification for you instead).



> > Thanks for your time reviewing the codes @andylin-hao. After removing current MUJOCO_EGL_DEVICE_ID settings, the egl rendering seems all on the first device? After that, the gpu memory is oom. image
> > Below is the right setting: image
> 
> Ah yes, sorry I wasn't very clear about this. My PR has not been merged yet, so you'll have to keep it for now. But you can write something like `os.environ["MUJOCO_EGL_DEVICE_ID"] = str(self._local_accelerator_id)`, which is a `Worker` member variable of the current assigned GPU. After my PR is merged, RLinf will automatically set this for every process so there's no need even for this. I'll let you know when the PR is merged (if your PR is merged first, I'll make the modification for you instead).

Get it! Thanks!😊😊😊

**Comment by zoeyuchao** [comment]
@chenkang455 Let's add Meta-world in the readme of the RLinf. The what's new part and the support matrix.

File: rlinf/algorithms/losses.py (hunk 1 above)
Comment (zanghz21) [reply rlinf/algorithms/losses.py:187]: This is correct.

File: rlinf/models/embodiment/openpi/__init__.py
Comment (zanghz21) [inline rlinf/models/embodiment/openpi/__init__.py]: This part looks great! I am considering whether we should use similar interfaces for other models. 
Actually, the function of this part is quite similar to  the EnvOutput().prepare_observations() in rlinf/data/io_struct.py

**Review by LiuZhihao2022 — APPROVED** [review]

**Comment by chenkang455** [comment]
@zoeyuchao Get it, I have added.

File: tests/e2e_tests/embodied/metaworld_50_ppo_openpi.yaml [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+rollout:
+  group_name: "RolloutGroup"
+  channel:
+    name: ${env.channel.name}
+    queue_name: "action_buffer"
+    queue_size: 0
+  mode: "colocate"
+  generation_backend: "huggingface"
+  model_dir: "/mnt/mnt/public/chenkang/openpi-main/checkpoints/sft/pi0_metaworld/metaworld_32_5k/4999"
```
Comment (chenkang455) [inline tests/e2e_tests/embodied/metaworld_50_ppo_openpi.yaml]: How should I set this CI path?

Comment (zoeyuchao) [reply tests/e2e_tests/embodied/metaworld_50_ppo_openpi.yaml]: @andylin-hao maybe Bro Lin can help u！

Comment (andylin-hao) [reply tests/e2e_tests/embodied/metaworld_50_ppo_openpi.yaml]: @chenkang455 I need to download the model to the CI machines first. Might take several hours. I'll let you know when it's done. For now you can write this path: `/workspace/dataset/RLinf-Pi0-MetaWorld-SFT`.

Comment (andylin-hao) [reply tests/e2e_tests/embodied/metaworld_50_ppo_openpi.yaml]: @chenkang455 Do you need me to upload this model to HuggingFace as well?

Comment (chenkang455) [reply tests/e2e_tests/embodied/metaworld_50_ppo_openpi.yaml]: @andylin-hao The checkpoint is saved on the J2 machine. Thanks for that 😊😊😊
```
model_dir: "/mnt/mnt/public/chenkang/openpi-main/checkpoints/sft/pi0_metaworld/metaworld_32_5k/4999"
```
And my current path is set as `pi0_metaworld` following `pi0_libero`
```
/workspace/dataset/pi0_metaworld
```

**Comment by andylin-hao** [comment]
Hi @chenkang455, I've made some necessary updates to the project dependency file to include MetaWorld, and enable CI test for MetaWorld in `.github/workflow`. I've made a PR to your branch here: https://github.com/chenkang455/RLinf/pull/1. Could you confirm and merge it into your branch?

**Comment by chenkang455** [comment]
@andylin-hao Of course, much appreciated. 🎉

**Comment by andylin-hao** [comment]
> @andylin-hao Of course, much appreciated. 🎉

Thanks. BTW, I'm still updating the CI machine to include MetaWorld. So the CI test is expected to fail. Don't worry, I'll rerun it once the machine update is complete.

**Comment by andylin-hao** [comment]
@chenkang455 After some testing, I found some issues with the dependencies in my previous PR. I made another one to fix it https://github.com/chenkang455/RLinf/pull/2. Now all checks have passed with the fix. Could you merge this one as well?

**Comment by chenkang455** [comment]
> @chenkang455 After some testing, I found some issues with the dependencies in my previous PR. I made another one to fix it [chenkang455#2](https://github.com/chenkang455/RLinf/pull/2). Now all checks have passed with the fix. Could you merge this one as well?

Get it, thanks for your check.

File: docs/source-en/rst_source/examples/metaworld.rst [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+
+2. **GRPO (Group Relative Policy Optimization)**
+
+   - For every state / prompt the policy generates *G* independent actions
+
+   - Compute the advantage of each action by subtracting the group's mean reward
+
+
+Dependency Installation
+--------------
```
Comment (QuanluZhang) [inline docs/source-en/rst_source/examples/metaworld.rst]: "-----" should have the same length as the title characters.

File: docs/source-en/rst_source/examples/metaworld.rst [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+      component_placement:
+         env: 0-1
+         rollout: 2-5
+         actor: 6-7
+
+You can also reconfigure the layout to achieve full separation,
+where env, rollout, and actor components each use their own GPUs with no
+interference, eliminating the need for offloading functionality.
+
+--------------
```
Comment (QuanluZhang) [inline docs/source-en/rst_source/examples/metaworld.rst]: what is this line used for?

File: docs/source-en/rst_source/examples/metaworld.rst [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+**2. Configuration Files**
+
+- π\ :sub:`0`\ + PPO:
+  ``examples/embodiment/config/metaworld_50_ppo_openpi.yaml``
+
+--------------
```
Comment (QuanluZhang) [inline docs/source-en/rst_source/examples/metaworld.rst]: this one too

**Comment by chenkang455** [comment]
@QuanluZhang Thanks for correcting them, I have updated the modified version in https://github.com/RLinf/RLinf/pull/242/commits/06e2f0d48d4ecffdb27d98463538e7ea25401994

File: rlinf/envs/metaworld/metaworld_env.py [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+                env_name = param["env_name"]
+                env = gym.make(
+                    "Meta-World/MT1",
+                    env_name=env_name,
+                    render_mode="rgb_array",
+                    camera_id=2,
+                    disable_env_checker=True,
+                )
+                # Set camera position to align with sft
+                env.env.env.env.env.env.env.model.cam_pos[2] = [0.75, 0.075, 0.7]
```
Comment (XuS1994) [inline rlinf/envs/metaworld/metaworld_env.py:111]: too many `env` here. What are they meaning? Can decrease?

File: rlinf/envs/metaworld/metaworld_env.py [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+                start_pivot = end_pivot
+
+        return np.array(task_ids), np.array(trial_ids)
+
+    @property
+    def elapsed_steps(self):
+        return self._elapsed_steps
+
+    @property
+    def info_logging_keys(self):
```
Comment (XuS1994) [inline rlinf/envs/metaworld/metaworld_env.py:198]: Seem not used?

File: rlinf/envs/metaworld/metaworld_env.py [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+            env_idx = np.arange(self.num_envs)
+
+        if reset_state_ids is None:
+            num_reset_states = len(env_idx)
+            reset_state_ids = self._get_random_reset_state_ids(num_reset_states)
+
+        self._reconfigure(reset_state_ids, env_idx)
+
+        if self.use_async_vector_env:
+            for _ in range(15):
```
Comment (XuS1994) [inline rlinf/envs/metaworld/metaworld_env.py]: Why use 15 here? Please replace with a environmental variables like `RESET_STEP`

File: rlinf/envs/metaworld/metaworld_env.py (hunk 9 above)
Comment (chenkang455) [reply rlinf/envs/metaworld/metaworld_env.py:111]: The environment has many wrappers so we have to get the internal environment through many envs, they can not be deleted.

File: rlinf/envs/metaworld/metaworld_env.py (hunk 10 above)
Comment (chenkang455) [reply rlinf/envs/metaworld/metaworld_env.py:198]: This code is borrowed from the libero_env.py?

File: rlinf/envs/metaworld/metaworld_env.py (hunk 11 above)
Comment (chenkang455) [reply rlinf/envs/metaworld/metaworld_env.py]: Get it!
